File: app/src/main/java/pro/anton/averin/networking/testrest/presenters/RequestPresenter.java
package pro.anton.averin.networking.testrest.presenters;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import okhttp3.Response;
import pro.anton.averin.networking.testrest.data.Repository;
import pro.anton.averin.networking.testrest.data.models.Request;
import pro.anton.averin.networking.testrest.rx.ResolvedObserver;
import pro.anton.averin.networking.testrest.rx.RxSchedulers;
import pro.anton.averin.networking.testrest.rx.events.DimBackgroundEvent;
import pro.anton.averin.networking.testrest.rx.events.FabClickedEvent;
import pro.anton.averin.networking.testrest.rx.events.UndimBackgroundEvent;
import pro.anton.averin.networking.testrest.utils.LLogger;

public class RequestPresenter extends RxBusPresenter<RequestView> {

    @Inject
    Repository repository;
    @Inject
    RxSchedulers schedulers;
    @Inject
    LLogger llogger;

    @Inject
    public RequestPresenter() {
    }

    public void onUseFileCheckboxChecked(boolean isChecked) {
        getView().cleanPostBody();
        if (isChecked) {
            getView().showPickFileButton();
            getView().setPostBodyFileHint();
        } else {
            getView().hidePickFileButton();
            getView().setPostBodyDefaultHint();
        }
    }

    public void onPickFileButtonClicked() {
        getView().showFileChooser();
    }

    public void onAddQueryButtonClicked() {
        getView().showAddQueryPopup();
    }

    public void onAddHeadersButtonClicked() {
        getView().showAddHeaderPopup();
    }

    public void onSendClicked(Request request) {
        if (request.isValid()) {
            repository.sendRequest(request).observeOn(schedulers.androidMainThread()).subscribe(new ResolvedObserver<Response>(
                                                                                                        getView().getUiResolution()) {
                @Override
                public void onCompleted() {
                    super.onCompleted();
                    view.navigateToResponseScreen();
                }

                @Override
                public void onNext(Response s) {
                    llogger.log(this, s.message());
                }
            });
        } else {
            getView().focusBaseUrl();
        }
    }

    public void onSaveItemClicked() {
        view.navigateToManagerScreenForSave();
    }

    public void onManagerItemClicked() {
        view.navigateToManagerScreen();
    }

    public void onClearItemClicked() {
        getView().clearFields();
    }

    public void onPostMethodSelected() {
        getView().showPostLayout();
    }

    public void onPostMethodUnselected() {
        getView().hidePostLayout();
    }

    public void addQueryPopupDismissed() {
        requestUnDimBackground();
    }

    public void requestUnDimBackground() {
        getGlobalBus().post(new UndimBackgroundEvent());
    }

    public void requestDimBackground() {
        getGlobalBus().post(new DimBackgroundEvent());
    }

    @Override
    protected void onEvent(@NotNull final Object event) {
        if (event instanceof FabClickedEvent && getVisible()) {
            onSendClicked(getView().getRequest());
        }
    }
}

File: app/src/main/java/pro/anton/averin/networking/testrest/presenters/RequestPresenter.kt
package pro.anton.averin.networking.testrest.presenters

import okhttp3.Response
import pro.anton.averin.networking.testrest.data.Repository
import pro.anton.averin.networking.testrest.data.models.Request
import pro.anton.averin.networking.testrest.rx.ResolvedObserver
import pro.anton.averin.networking.testrest.rx.RxSchedulers
import pro.anton.averin.networking.testrest.rx.events.DimBackgroundEvent
import pro.anton.averin.networking.testrest.rx.events.FabClickedEvent
import pro.anton.averin.networking.testrest.rx.events.UndimBackgroundEvent
import javax.inject.Inject

class RequestPresenter
@Inject
constructor() : RxBusPresenter<RequestView>() {

    @Inject
    lateinit var repository: Repository
    @Inject
    lateinit var schedulers: RxSchedulers

    fun onUseFileCheckboxChecked(isChecked: Boolean) {
        view.cleanPostBody()
        if (isChecked) {
            view.showPickFileButton()
            view.setPostBodyFileHint()
        } else {
            view.hidePickFileButton()
            view.setPostBodyDefaultHint()
        }
    }

    fun onPickFileButtonClicked() {
        view.showFileChooser()
    }

    fun onAddQueryButtonClicked() {
        view.showAddQueryPopup()
    }

    fun onAddHeadersButtonClicked() {
        view.showAddHeaderPopup()
    }

    fun onSendClicked(request: Request) {
        if (request.isValid) {
            repository.sendRequest(request).observeOn(schedulers.androidMainThread()).subscribe(object : ResolvedObserver<Response>(
                    view.uiResolution) {
                override fun onCompleted() {
                    super.onCompleted()
                    view.navigateToResponseScreen()
                }

                override fun onNext(s: Response) {
                    llogger.log(this, s.message())
                }
            })
        } else {
            view.focusBaseUrl()
        }
    }

    fun onSaveItemClicked() {
        view.navigateToManagerScreenForSave()
    }

    fun onManagerItemClicked() {
        view.navigateToManagerScreen()
    }

    fun onClearItemClicked() {
        view.clearFields()
    }

    fun onPostMethodSelected() {
        view.showPostLayout()
    }

    fun onPostMethodUnselected() {
        view.hidePostLayout()
    }

    fun addQueryPopupDismissed() {
        requestUnDimBackground()
    }

    fun requestUnDimBackground() {
        globalBus.post(UndimBackgroundEvent())
    }

    fun requestDimBackground() {
        globalBus.post(DimBackgroundEvent())
    }

    override fun onEvent(event: Any) {
        if (event is FabClickedEvent && visible) {
            onSendClicked(view.request)
        }
    }
}
